L1Trigger/TrackerTFP/src/HoughTransform.cc

```diff
@@ -0,0 +1,200 @@
+#include "L1Trigger/TrackerTFP/interface/HoughTransform.h"
+
+#include <numeric>
+#include <algorithm>
+#include <iterator>
+#include <deque>
+#include <vector>
+#include <set>
+#include <utility>
+#include <cmath>
+
+using namespace std;
+using namespace edm;
+using namespace trackerDTC;
+
+namespace trackerTFP {
+
+  HoughTransform::HoughTransform(const ParameterSet& iConfig, const Setup* setup, const DataFormats* dataFormats, int region) :
+    enableTruncation_(iConfig.getParameter<bool>("EnableTruncation")),
+    setup_(setup),
+    dataFormats_(dataFormats),
+    inv2R_(dataFormats_->format(Variable::inv2R, Process::ht)),
+    phiT_(dataFormats_->format(Variable::phiT, Process::ht)),
+    region_(region),
+    input_(dataFormats_->numChannel(Process::ht), vector<deque<StubGP*>>(dataFormats_->numChannel(Process::gp)))
+  {}
+
+  // read in and organize input product
+  void HoughTransform::consume(const TTDTC::Streams& streams) {
+    const int offset = region_ * dataFormats_->numChannel(Process::gp);
+    auto validFrame = [](int& sum, const TTDTC::Frame& frame){ return sum += frame.first.isNonnull() ? 1 : 0; };
+    int nStubsGP(0);
+    for (int sector = 0; sector < dataFormats_->numChannel(Process::gp); sector++) {
+      const TTDTC::Stream& stream = streams[offset + sector];
+      nStubsGP += accumulate(stream.begin(), stream.end(), 0, validFrame);
+    }
+    stubsGP_.reserve(nStubsGP);
+    for (int sector = 0; sector < dataFormats_->numChannel(Process::gp); sector++) {
+      const int sectorPhi = sector % setup_->numSectorsPhi();
+      const int sectorEta = sector / setup_->numSectorsPhi();
+      for (const TTDTC::Frame& frame : streams[offset + sector]) {
+        StubGP* stub = nullptr;
+        if (frame.first.isNonnull()) {
+          stubsGP_.emplace_back(frame, dataFormats_, sectorPhi, sectorEta);
```

Comment:
Add comment such as "Store input stubs in array, so rest of HT algo can work with pointers to them (saves CPU)".

Reply:
done